### DISCUSSION TITLE
The timing of publishing for domain events using Mediatr

### DISCUSSION THREAD

**Discussion by thatdotnetguy** [LuckyPennySoftware/MediatR#1028] [Q&A]
I can't see why doing a savechanges interceptor to fire off the domain events is helpful or what it adds to an architecture as opposed to just publishing (eg await _mediator.Publish(domainEvent)) them from within the handler?

![image](https://github.com/jbogard/MediatR/assets/63023163/f70b1edd-4566-4236-ac5b-db65d98f882c)
![image](https://github.com/jbogard/MediatR/assets/63023163/677118a1-6e4e-4f66-bd89-52b2b8feea23)

**Comment by ardalis** [comment]
It's helpful because it ensure that if you're doing something with external side effects, that only occurs after the save has completed successfully.

This way you don't send an email thanking the customer for their order and then fail to save the order, for example.

I cover pre and post persistence events in episodes on weeklydevtips.com if you want more background.

**Reply by thatdotnetguy** [reply]
Hi @ardalis I'm not sure I follow. If we use your example about email... the above code would send the email (send an event) and then attempt to save.

Maybe you're thinking about publishing before SavedChangesAsync instead of before SavingChangesAsync is called?

**Reply by ardalis** [reply]
Good point. I was replying from my phone (then and now) and hadn't looked closely at the images. If you look at the code in my Clean Architecture sample (which this code may have borrowed from) you'll see that I do it in reverse order, dispatching only after successfully saving. Which is what I recommend in most cases.

**Reply by thatdotnetguy** [reply]
Cheers Steve.... we were confused... you seem like quite a heavy hitter so thought you might have responded quickly :)

I'll check out your sample for sure.... this sample was from someone else so I'm not sure what they're on about :)
